---
title: "Variant Annotation Workshop with FunciVAR, StateHub and MotifBreakR"
author: "Simon G. Coetzee & Dennis J. Hazelett"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{FunciVar, MotifBreakR and StatePaintR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# FuncVAR: Annotation and functional enrichment of variant sets

In this vignette we will learn to take a set of variants, arbitrarily defined as any set of GenomicRanges, overlap with multiomics datasets and then calculate enrichment against an appropriately chosen background set.

## Setup

First we need to import the packages we'll be using. FunciVar can be accessed directly from our github repositories via biocLite. MotifbreakR, GenomicRanges and VariantAnnotation packages are available through bioconductor:

```{r}
source("https://bioconductor.org/biocLite.R")
biocLite("Simon-Coetzee/StatePaintR", suppressUpdates = TRUE, dependencies = TRUE)
biocLite("Simon-Coetzee/funciVar", suppressUpdates = TRUE, dependencies = TRUE)
biocLite("motifbreakR")
library(StatePaintR)
library(funciVar)
library(motifbreakR)
library(GenomicRanges)
library(VariantAnnotation)
```

## Getting annotations

Annotations can come from any source, the key requirement be that they be coerced into GRanges format for use by the funciVar package. For our purposes, we have populated a directory called "segmentations" with a set of bed files representing all the cell types in the Roadmap epigenomics project. Each segmentation file contains a set of chromatin state calls made by our StatePaintR package (covered in a later section).

FunciVar imports these files efficiently by pointing a function called "GetSegmentations" directly at this directory. Then it's just a matter of renaming the state abbreviations to easily recognized, human readable form.

```{r}
#### MODIFY THIS TO DOWNLOAD LIVE OR MAKE SURE TRACKS ARE INSIDE PACKAGE
segmentation.files <- list.files("/common/bioc_workflow/segmentations/", full.names = TRUE)
encode.segmenations <- GetSegmentations(segmentation.files, genome = "hg19")
encode.segmenations
encode.segmenations <- unlist(encode.segmenations)
encode.segmenations.granular <- encode.segmenations
mcols(encode.segmenations[encode.segmenations$state %in% c("EAR", "AR"), ])$state <- "Active Enhancer"
mcols(encode.segmenations[encode.segmenations$state %in% c("EPR", "EWR"), ])$state <- "Weak Enhancer"
mcols(encode.segmenations[encode.segmenations$state %in% c("PAR"), ])$state <- "Active Promoter"
mcols(encode.segmenations[encode.segmenations$state %in% c("PPR", "PPWR", "PWR"), ])$state <- "Weak Promoter"
mcols(encode.segmenations[encode.segmenations$state %in% c("CTCF"), ])$state <- "CTCF"
mcols(encode.segmenations[encode.segmenations$state %in% c("HET", "SCR", "TRS"), ])$state <- "Other"
```

## Creating the foreground variant set

There are two potential use cases for funciVar. In the first case, a foreground set is predefined and nothing needs to be done outside of importing to R and coercing into GRanges. Such foreground sets could include any kind of variant; common population polymorphisms or differentially methylated CpGs for example. 

In the second case, especially common for primary GWAS studies, a single top "hit" has been identified and a set of variants known to be in linkage disequilibrium (LD) should be calculated. We will cover the latter case. The LD calculations can be used to further subdivide regional variants into foreground and background based on some LD threshold (typically at r^2 = 0.8). In other instances, more creative methods of obtaining a background set might be warranted.

We will use 1000 genomes data for our LD calculations, but in principle we could apply this to any population stored in VCF. First we need to identify the file containing the 1000 and assign all the patient metadata into a variable called "my.samples".

```{r}
my.samples <- read.delim("http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel", stringsAsFactors = FALSE)
my.samples <- my.samples[, c(1,2,3,4)]
```

Next we will identify the file with chromosome 10 variants and define the index region as a filter for the foreground variant set. For simplicity's sake, we are testing a single region, but we could easily analyze multiple ranges. For this example, we are using a GWAS variant linked to breast cancer on chromosome 10.

```{r}
chr10.remote.vcf <- "ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr10.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"
index10q26 <- "rs2981579"
### PMID: 23535729
#### I THINK EBI is rate limiting requests heavily, we must either host these two chromosomes on our AWS or package with these rda files.
#save(vcf10q26, file = "vcf10q26.rda")
#load("vcf10q26.rda")
pos10q26 <- promoters(as("10:123337335-123337335", "GRanges"), upstream = 500000, downstream = 500000)
```

### Calculation of LD

Using the 1MB window that we defined for the index variant, we capture the relevant variants from the VCF file for our LD calculations.

```{r}
vcf10q26 <- GetVariantsInWindow(file = chr10.remote.vcf, position = pos10q26)
```

To ensure that we make our LD calculation from the correct population structure, we set the population using metadata "my.samples".

```{r}
vcf10q26 <- SetPopulation(vcf10q26, sample_sheet = my.samples)
```

Now we calculate LD structure relative to our index variant in the european population background.

```{r}
vcf10q26snps <- CalcLD(vcf10q26, index10q26, "EUR")
```

Next we use the "SplitVcfLd" as a convenience function to define our foreground and background sets based on LD. This is a sensible choice since other variants in the same region are likely to have similar properties relative to genomic features like gene Density, gc bias, epigenetic marks etc. As an alternative, we might consider using a resource like SNPsnap to match these properties genome wide.

```{r}
vcf10q26snps <- SplitVcfLd(vcf10q26snps)
```

## Functional enrichment by chromatin state

In the next part of our workflow, we want to determine whether the index SNP and its LD proxies, which we defined to be in the "foreground", are enriched in specific cellular features. As alluded to earlier, the FunciVar package could be used to carry out any analagous operation to find enrichment of one set of features in another. Here we plan to use chromatin state calls from 119 human tissues in Roadmap, which we've imported at the beginning of our session. We specify that we are using this type of annotation with the "feature.type" argument (alternative "biofeatures" argument can be used for single element features, _e.g._ a bed file with K27ac peaks). Using the "CalculateEnrichment" function we generate an object containing summary enrichment statistics.

### Enrichment calculations

FunciVar by default calculates a likelihood based on the beta-binomial distribution, returning a 95% credible interval (optionally set by the "CI" argument) for the range of differences between the two populations of variants (_i.e._ foreground and background). Specifically it calculates a distribution of true enrichment (as probability of overlap) for both sets of variants in the genomic features based on the observed number of overlaps:

```{latex}
\[
\theta_{fg} ~ beta(S_{fg} + a, N_{fg} + b)
\theta_{bg} ~ beta(S_{bg} + a, N_{bg} + b)
\]
```

FunciVar uses an uninformative Jeffreys prior[^ Jeffreys] to compare the two distributions directly by subtracting permuted samples to obtain the distribution of differences. The prior can be overidden in special cases (see documentation).

[^ Jeffreys] A note about Jeffreys prior and links here.

To perform these operations with funciVar, we call the "CalculateEnrichment" function.

```{r}
enrich10q26 <- CalculateEnrichment(variants = vcf10q26snps, features = encode.segmenations, feature.type = "segmentations", CI = 0.8, strict.subset = TRUE, return.overlaps = TRUE)
```

Now we'll extract the enrichment stats and alter the sample labels to remove the file suffix inherited from the genomic feature bed files.

```{r}
enrich10q26.enrich <- enrich10q26$enrichment
enrich10q26.enrich$sample <- stringr::str_replace(enrich10q26.enrich$sample, "\\.\\d*mark.segmentation.bed", "")
PlotEnrichment(enrich10q26.enrich, value = "difference", block1 = "state", color.by = "sample", ncol = 6)
```

### Annotations of interest

While there is a tendency for SNPs in LD with each other to be enriched in some cellular interest, it is probable that only a subset of them are of functional interest. In our case, we would like to examine the hypothesis that one of our putative functional SNPs disrupts a transcription factor binding site. 

Since our index SNP was found in association with breast cancer, let's choose the MCF7 cell line from Roadmap to identify proxies that overlap with active enhancers. Segmentation features live in the overlaps slot of the enrichment object from the previous steps.

```{r}
snps.10q26.enhancers <- enrich10q26$overlaps$`Active Enhancer`$foregound.overlaps[enrich10q26$overlaps$`Active Enhancer`$foregound.overlaps$mcf.7.12mark.segmentation.bed > 0, ]
```

# MotifBreakR analysis of TF binding disruption

Now that we have a SNP of interest from breast tissue enhancer, let's work with the motifBreakR package to determine whether a putative transcription factor binding site (TFBS) is disrupted. MotifBreakR allows us to do this with nothing more than the rsID or a location and allele information. It enables us to use any genome housed in bioconductor BSgenome packages and any set of motifs curated in the MotifDB package (plus some custom databases in MotifDB format that are included with the MotifBreakR installation)[^Motifbreakr Manuscript].

[^Motifbreakr Manuscript] see the manuscript for details.

## Setup

We need to import some version of the human genome and a compatible set of variants from Bioconductor.

```{r}
library(BSgenome.Hsapiens.UCSC.hg19)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
```

We need to grab the SNP variant objects out of SNPlocs package that correspond to our enhancer proxies.

```{r}
snps.mb.10q26.enhancers <- snps.from.rsid(rsid = names(snps.10q26.enhancers),
                                          dbSNP = SNPlocs.Hsapiens.dbSNP144.GRCh37,
                                          search.genome = BSgenome.Hsapiens.UCSC.hg19)
```

## Analysis

Now a simple call to the motifbreakR function completes the analysis. Here, we specify that we want to filter our results by p-value (rather than scaled motif scores), we'll set that p-value threshold at 10^-4, using information content (set argument *method = "ic"*) and a uniform prior on the GC content.

```{r}
results.10q26.enhancers <- motifbreakR(snpList = snps.mb.10q26.enhancers, filterp = TRUE,
                                       pwmList = hocomoco,
                                       threshold = 1e-4,
                                       method = "ic",
                                       bkg = c(A=0.25, C=0.25, G=0.25, T=0.25),
                                       BPPARAM = BiocParallel::bpparam())
```

MotifbreakR does a quick estimate of the p-value threshold initially. To obtain accurate p-values, we run "calculatePvalue" in a seperate step (beware this can be slow!).

```{r}
pvalue.results.10q26.enhancers <- calculatePvalue(results = results.10q26.enhancers)
```

## Visualization

MotifbreakR includes a plotting function to visualize the SNP in its genomic context, aligned with the motif logos it is purported to alter.

```{r}
plotMB(pvalue.results.10q26.enhancers, "rs34354213", effect = "strong")
```

# Second example


```{r}
chr1.remote.vcf <- "ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"
index1q21.1 <- "rs12405132"
### PMID: 25751625
### I THINK EBI is rate limiting requests heavily, we must either host these two chromosomes on our AWS or package with these rda files.
pos1q21.1 <- promoters(as("1:145644984-145644984", "GRanges"), upstream = 500000, downstream = 500000)
```
```{r}
vcf1q21.1 <- GetVariantsInWindow(file = chr1.remote.vcf, position = pos1q21.1)
vcf1q21.1 <- SetPopulation(vcf1q21.1, sample_sheet = my.samples)
vcf1q21.1snps <- CalcLD(vcf1q21.1, index1q21.1, "EUR")
vcf1q21.1snps <- SplitVcfLd(vcf1q21.1snps)
enrich1q21.1 <- CalculateEnrichment(variants = vcf1q21.1snps, features = encode.segmenations, feature.type = "segmentations", strict.subset = TRUE, return.overlaps = TRUE)
```
```{r}
enrich1q21.1.enrich <- enrich1q21.1$enrichment
enrich1q21.1.enrich$sample <- stringr::str_replace(enrich1q21.1.enrich$sample, "\\.\\d*mark.segmentation.bed", "")
PlotEnrichment(enrich1q21.1.enrich, value = "difference", block1 = "state", color.by = "sample", ncol = 6)
```
```{r}
snps.1q21.1.enhancers <- enrich1q21.1$overlaps$`Active Enhancer`$foregound.overlaps[enrich1q21.1$overlaps$`Active Enhancer`$foregound.overlaps$mcf.7.12mark.segmentation.bed > 0, ]
snps.mb.1q21.1.enhancers <- snps.from.rsid(rsid = names(snps.1q21.1.enhancers),
                                          dbSNP = SNPlocs.Hsapiens.dbSNP144.GRCh37,
                                          search.genome = BSgenome.Hsapiens.UCSC.hg19)
```
```{r}
results.1q21.1.enhancers <- motifbreakR(snpList = snps.mb.1q21.1.enhancers, filterp = TRUE,
                                       pwmList = hocomoco,
                                       threshold = 1e-4,
                                       method = "ic",
                                       bkg = c(A=0.25, C=0.25, G=0.25, T=0.25),
                                       BPPARAM = BiocParallel::bpparam())
pvalue.results.1q21.1.enhancers <- calculatePvalue(results = results.1q21.1.enhancers)
```
```{r}
plotMB(pvalue.results.1q21.1.enhancers, "rs2040086")
plotMB(pvalue.results.1q21.1.enhancers, "rs2040085")
```
