---
title: "Vignette Title"
author: "Vignette Author"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r}
biocLite("Simon-Coetzee/StatePaintR", suppressUpdates = TRUE, dependencies = TRUE)
biocLite("Simon-Coetzee/funciVar", suppressUpdates = TRUE, dependencies = TRUE)
library(StatePaintR)
library(funciVar)
library(motifbreakR)
library(GenomicRanges)
library(VariantAnnotation)
```
```{r}
#### MODIFY THIS TO DOWNLOAD LIVE OR MAKE SURE TRACKS ARE INSIDE PACKAGE
segmentation.files <- list.files("/common/bioc_workflow/segmentations/", full.names = TRUE)
encode.segmenations <- GetSegmentations(segmentation.files, genome = "hg19")
encode.segmenations
encode.segmenations <- unlist(encode.segmenations)
encode.segmenations.granular <- encode.segmenations
mcols(encode.segmenations[encode.segmenations$state %in% c("EAR", "AR"), ])$state <- "Active Enhancer"
mcols(encode.segmenations[encode.segmenations$state %in% c("EPR", "EWR"), ])$state <- "Weak Enhancer"
mcols(encode.segmenations[encode.segmenations$state %in% c("PAR"), ])$state <- "Active Promoter"
mcols(encode.segmenations[encode.segmenations$state %in% c("PPR", "PPWR", "PWR"), ])$state <- "Weak Promoter"
mcols(encode.segmenations[encode.segmenations$state %in% c("CTCF"), ])$state <- "CTCF"
mcols(encode.segmenations[encode.segmenations$state %in% c("HET", "SCR", "TRS"), ])$state <- "Other"
```
```{r}
my.samples <- read.delim("http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel", stringsAsFactors = FALSE)
my.samples <- my.samples[, c(1,2,3,4)]
```
```{r}
chr10.remote.vcf <- "ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr10.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"
index10q26 <- "rs2981579"
### PMID: 23535729
#### I THINK EBI is rate limiting requests heavily, we must either host these two chromosomes on our AWS or package with these rda files.
#save(vcf10q26, file = "vcf10q26.rda")
#load("vcf10q26.rda")
pos10q26 <- promoters(as("10:123337335-123337335", "GRanges"), upstream = 500000, downstream = 500000)
```
```{r}
vcf10q26 <- GetVariantsInWindow(file = chr10.remote.vcf, position = pos10q26)
```
```{r}
vcf10q26 <- SetPopulation(vcf10q26, sample_sheet = my.samples)
```
```{r}
vcf10q26snps <- CalcLD(vcf10q26, index10q26, "EUR")
```
```{r}
vcf10q26snps <- SplitVcfLd(vcf10q26snps)
```
```{r}
enrich10q26 <- CalculateEnrichment(variants = vcf10q26snps, features = encode.segmenations, feature.type = "segmentations", CI = 0.8, strict.subset = TRUE, return.overlaps = TRUE)
```
```{r}
enrich10q26.enrich <- enrich10q26$enrichment
enrich10q26.enrich$sample <- stringr::str_replace(enrich10q26.enrich$sample, "\\.\\d*mark.segmentation.bed", "")
PlotEnrichment(enrich10q26.enrich, value = "difference", block1 = "state", color.by = "sample", ncol = 6)
```
```{r}
snps.10q26.enhancers <- enrich10q26$overlaps$`Active Enhancer`$foregound.overlaps[enrich10q26$overlaps$`Active Enhancer`$foregound.overlaps$mcf.7.12mark.segmentation.bed > 0, ]
```
```{r}
library(BSgenome.Hsapiens.UCSC.hg19)
library(SNPlocs.Hsapiens.dbSNP144.GRCh37)
```
```{r}
snps.mb.10q26.enhancers <- snps.from.rsid(rsid = names(snps.10q26.enhancers),
                                          dbSNP = SNPlocs.Hsapiens.dbSNP144.GRCh37,
                                          search.genome = BSgenome.Hsapiens.UCSC.hg19)
```
```{r}
results.10q26.enhancers <- motifbreakR(snpList = snps.mb.10q26.enhancers, filterp = TRUE,
                                       pwmList = hocomoco,
                                       threshold = 1e-4,
                                       method = "ic",
                                       bkg = c(A=0.25, C=0.25, G=0.25, T=0.25),
                                       BPPARAM = BiocParallel::bpparam())
```
```{r}
pvalue.results.10q26.enhancers <- calculatePvalue(results = results.10q26.enhancers)
```
```{r}
plotMB(pvalue.results.10q26.enhancers, "rs34354213", effect = "strong")
```
```{r}
chr1.remote.vcf <- "ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr1.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz"
index1q21.1 <- "rs12405132"
### PMID: 25751625
### I THINK EBI is rate limiting requests heavily, we must either host these two chromosomes on our AWS or package with these rda files.
pos1q21.1 <- promoters(as("1:145644984-145644984", "GRanges"), upstream = 500000, downstream = 500000)
```
```{r}
vcf1q21.1 <- GetVariantsInWindow(file = chr1.remote.vcf, position = pos1q21.1)
vcf1q21.1 <- SetPopulation(vcf1q21.1, sample_sheet = my.samples)
vcf1q21.1snps <- CalcLD(vcf1q21.1, index1q21.1, "EUR")
vcf1q21.1snps <- SplitVcfLd(vcf1q21.1snps)
enrich1q21.1 <- CalculateEnrichment(variants = vcf1q21.1snps, features = encode.segmenations, feature.type = "segmentations", strict.subset = TRUE, return.overlaps = TRUE)
```
```{r}
enrich1q21.1.enrich <- enrich1q21.1$enrichment
enrich1q21.1.enrich$sample <- stringr::str_replace(enrich1q21.1.enrich$sample, "\\.\\d*mark.segmentation.bed", "")
PlotEnrichment(enrich1q21.1.enrich, value = "difference", block1 = "state", color.by = "sample", ncol = 6)
```
```{r}
snps.1q21.1.enhancers <- enrich1q21.1$overlaps$`Active Enhancer`$foregound.overlaps[enrich1q21.1$overlaps$`Active Enhancer`$foregound.overlaps$mcf.7.12mark.segmentation.bed > 0, ]
snps.mb.1q21.1.enhancers <- snps.from.rsid(rsid = names(snps.1q21.1.enhancers),
                                          dbSNP = SNPlocs.Hsapiens.dbSNP144.GRCh37,
                                          search.genome = BSgenome.Hsapiens.UCSC.hg19)
```
```{r}
results.1q21.1.enhancers <- motifbreakR(snpList = snps.mb.1q21.1.enhancers, filterp = TRUE,
                                       pwmList = hocomoco,
                                       threshold = 1e-4,
                                       method = "ic",
                                       bkg = c(A=0.25, C=0.25, G=0.25, T=0.25),
                                       BPPARAM = BiocParallel::bpparam())
pvalue.results.1q21.1.enhancers <- calculatePvalue(results = results.1q21.1.enhancers)
```
```{r}
plotMB(pvalue.results.1q21.1.enhancers, "rs2040086")
plotMB(pvalue.results.1q21.1.enhancers, "rs2040085")
```
